broker/pkg/coordinator/grpcserver.go

```diff
@@ -92,15 +94,56 @@ func (s *GRPCServer) TimelineConsumerHandshake(ctx context.Context, req *grpc.Ti
 	return builder, err
 }
 
-func (s *GRPCServer) TimelineConsume(req *grpc.TimelineConsumeRequest, stream grpc.Broker_TimelineConsumeServer) error {
-	pipeline, err := s.listeners.ConsumerConnected(stream.Context(), string(req.SessionID()))
+func (s *GRPCServer) TimelineConsume(stream grpc.Broker_TimelineConsumeServer) error {
+	var sessionID []byte
+	var sessionLock sync.RWMutex
+	sessionLock.Lock()
+	go func() {
+		var err error
+		var request *grpc.TimelineConsumerStatus
+		for err == nil {
+			select {
+			case <-stream.Context().Done():
+				return
+			default:
+				request, err = stream.Recv()
+				if request == nil { //empty status ?!?!?! TODO log this as a warning
+					continue
+				}
+				if err == io.EOF { //it means the stream batch is over
+					break
+				}
+				if err != nil {
+					_ = fmt.Errorf("grpc server TimelineCreateMessages client failed err=%s", err.Error())
+					break
+				}
+				if sessionID == nil {
+					sessionID = request.SessionID()
+					sessionLock.Unlock()
+				}
+
+				consumerStatus := protocol.NewConsumerStatus(sessionID)
+				consumerStatus.AvailableBufferSize = request.AvailableBufferSize()
+				consumerStatus.MaxBufferSize = request.MaxBufferSize()
+				s.listeners.ConsumerStatus(stream.Context(), consumerStatus)
+
+				if err != nil {
+					log.Fatalf("Update Consumer status failed, err=%s", err)
+				}
+			}
+		}
+	}()
+
+	sessionLock.Lock()
+	pipeline, err := s.listeners.ConsumerConnected(stream.Context(), string(sessionID))
 	if err != nil {
 		return err
 	}
 
 	defer func() {
-		s.listeners.ConsumerDisconnected(stream.Context(), string(req.SessionID()))
+		s.listeners.ConsumerDisconnected(stream.Context(), string(sessionID))
 	}()
+	sessionLock.Unlock()
```

Comment:
if err will occur at line 140, this will never happen, and will remain locked. I suggest putting a defer after the Lock.

Reply:
I replaced this with a channel sync mechanism